ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/editor/context/InputContextManager.java

```diff
@@ -208,30 +208,42 @@ public void resourceChanged(IResourceChangeEvent event) {
 		IResourceDelta delta = event.getDelta();
 
 		try {
+			List<IFile> added = new ArrayList<>();
+			List<IFile> removed = new ArrayList<>();
 			delta.accept(delta1 -> {
-				int kind = delta1.getKind();
 				IResource resource = delta1.getResource();
 				if (resource instanceof IFile) {
+					int kind = delta1.getKind();
 					if (kind == IResourceDelta.ADDED)
-						asyncStructureChanged((IFile) resource, true);
+						added.add((IFile) resource);
 					else if (kind == IResourceDelta.REMOVED)
-						asyncStructureChanged((IFile) resource, false);
+						removed.add((IFile) resource);
 					return false;
 				}
 				return true;
 			});
+			if (!added.isEmpty() || !removed.isEmpty()) {
+				asyncStructureChanged(added, removed);
+			}
 		} catch (CoreException e) {
 			PDEPlugin.logException(e);
 		}
 	}
 
-	private void asyncStructureChanged(final IFile file, final boolean added) {
+	private void asyncStructureChanged(List<IFile> added, List<IFile> removed) {
 		if (editor == null || editor.getEditorSite() == null)
 			return;
 		Shell shell = editor.getEditorSite().getShell();
 		Display display = shell != null ? shell.getDisplay() : Display.getDefault();
```

Comment:
Again, it was that way before...

But yes, the code is rather awful and  it strikes me that PlatformUI.getWorkbench().getDisplay() would be a much simpler way to get a display.   The existing logic doesn't check isDisposed on the Display, but I suppose that's not possible while processing a resource delta...

Reply:
Ed, dispose checks are possible from non UI thread, but if it is not disposed, it doesn't mean anything, because it has to be checked again at the task start time in UI thread.

Reply:
That makes you wonder how it's ever safe to call asyncExec because at any point the display can be disposed
```
public void asyncExec (Runnable runnable) {
	synchronized (Device.class) {
		if (isDisposed ()) error (SWT.ERROR_DEVICE_DISPOSED);
		synchronizer.asyncExec (runnable);
	}
}
```
But digressing into theoretical problems that can't really happen in practice won't serve us well, will it?   After all, once the display itself is disposed, the user won't see anything ever again.  So let's not go there.

For now I force pushed a change that is not error prone.  I'm not so convinced a UIJob will serve us well, because I don't think not updating the visible state will be good for the user...

Please comment on the newer changes if you still feel strongly about that.